src/backends/tensorflow/tensorflow_backend_tf.cc

```diff
@@ -399,35 +501,82 @@ ModelImpl::~ModelImpl()
   TRTISTF_IOListDelete(outputs_);
 }
 
+TRTISTF_Error*
+ModelImpl::MakeCallable(const tensorflow::CallableOptions& opts)
+{
+  if (has_callable_) {
+    session_->ReleaseCallable(callable_).IgnoreError();
+    has_callable_ = false;
+  }
+
+  RETURN_IF_TF_ERROR(session_->MakeCallable(opts, &callable_));
+  for (int idx = 0; idx < opts.fetch_size(); idx++) {
+    output_index_map_[opts.fetch(idx)] = idx;
+  }
+  has_callable_ = true;
+  return nullptr;
+}
+
 TRTISTF_Error*
 ModelImpl::Run(
     TRTISTF_TensorList* input_tensors,
     const std::vector<std::string>& output_names,
     TRTISTF_TensorList** output_tensors)
 {
-  std::vector<std::pair<std::string, tensorflow::Tensor>> tfinputs;
-
-  for (TRTISTF_TensorList* itr = input_tensors; itr != nullptr;
-       itr = itr->next_) {
-    if (itr->tensor_ != nullptr) {
-      TensorImpl* tensor = reinterpret_cast<TensorImpl*>(itr->tensor_);
-      tfinputs.emplace_back(
-          std::make_pair(tensor->Name(), std::move(tensor->TFTensor())));
+  // I/O needs to be prepared differently for callable
+  if (has_callable_) {
+    std::vector<tensorflow::Tensor> tfinputs;
+
+    for (TRTISTF_TensorList* itr = input_tensors; itr != nullptr;
+         itr = itr->next_) {
+      if (itr->tensor_ != nullptr) {
+        TensorImpl* tensor = reinterpret_cast<TensorImpl*>(itr->tensor_);
+        tfinputs.emplace_back(std::move(tensor->TFTensor()));
+      }
+    }
+    TRTISTF_TensorListDelete(input_tensors);
+
+    tensorflow::RunMetadata meta_data;
+    std::vector<tensorflow::Tensor> tfoutputs;
+    RETURN_IF_TF_ERROR(
+        session_->RunCallable(callable_, tfinputs, &tfoutputs, &meta_data));
+
+    *output_tensors = nullptr;
+    for (auto ri = output_names.rbegin(); ri != output_names.rend(); ++ri) {
+      const auto oidx = output_index_map_[*ri];
+      TRTISTF_Tensor* tensor = reinterpret_cast<TRTISTF_Tensor*>(
+          new TensorImpl(std::move(tfoutputs[oidx])));
+      *output_tensors = TRTISTF_TensorListNew(tensor, *output_tensors);
     }
-  }
-
-  TRTISTF_TensorListDelete(input_tensors);
-
-  std::vector<tensorflow::Tensor> tfoutputs;
-  RETURN_IF_TF_ERROR(session_->Run(tfinputs, output_names, {}, &tfoutputs));
 
-  *output_tensors = nullptr;
-  for (std::vector<tensorflow::Tensor>::reverse_iterator ri =
-           tfoutputs.rbegin();
-       ri != tfoutputs.rend(); ++ri) {
-    TRTISTF_Tensor* tensor =
-        reinterpret_cast<TRTISTF_Tensor*>(new TensorImpl(std::move(*ri)));
-    *output_tensors = TRTISTF_TensorListNew(tensor, *output_tensors);
+    // Have to sync device to ensure output tensors are ready,
+    // because the Callable is created with 'fetch_skip_sync = true'
+    // (false is not implemented).
+    device_->Sync();
```

Comment:
This is potentially very bad for performance of not just this model but other things running on the GPU. Does the Run return before the outputs are computed?

Reply:
That's possible.. I am not sure whether the outputs are actually ready when `RunCallable()` returns, the `Device::Sync()` call is suggested by the [Callable doc](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/protobuf/config.proto#L752), but I have also done some image classifications without the sync call and the client responses are consistent.

I can remove the sync and add on the test to verify the correctness on large payload (identity model with large input), if the output is consistent, then we can be confident to remove the sync call?

Reply:
I don't know... but I assume that after the run call the results are sitting somewhere on the GPU. Hopefully the callable doesn't move them around async to the actual Run call. Or does the callable make the Run async somehow?

Reply:
Removed, and added inline comment on why